src/rust/catnap/socket.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+//======================================================================================================================
+// Imports
+//======================================================================================================================
+
+use crate::runtime::fail::Fail;
+use ::std::net::SocketAddrV4;
+
+//======================================================================================================================
+// Structures
+//======================================================================================================================
+
+/// Encodes the state of a socket.
+#[derive(Copy, Clone, Debug)]
+enum SocketState {
+    /// A socket that is not bound.
+    NotBound,
+    /// A socket that is bound to a local address.
+    Bound,
+    /// A socket that is bound to a local address and is able to accept incoming connections.
+    Listening,
+    /// A socket that is bound to a local address and is accepting incoming connections.
+    Accepting,
+    /// A socket that is attempting to connect to a remote address.
+    Connecting,
+    /// A socket that is connected to a remote address.
+    Connected,
+    /// A socket that is closed.
+    Closed,
+}
+
+/// A socket.
+#[derive(Copy, Clone, Debug)]
+pub struct Socket {
+    /// The state of the socket.
+    state: SocketState,
+    /// The local address to which the socket is bound.
+    local: Option<SocketAddrV4>,
+    /// The remote address to which the socket is connected.
+    remote: Option<SocketAddrV4>,
+}
+
+//======================================================================================================================
+// Associated Functions
+//======================================================================================================================
+
+impl Socket {
+    /// Creates a new socket that is not bound to an address.
+    pub fn new() -> Self {
+        Self {
+            state: SocketState::NotBound,
+            local: None,
+            remote: None,
+        }
+    }
+
+    /// Constructs from [self] a socket that is bound to the `local` address.
+    pub fn bind(&self, local: SocketAddrV4) -> Result<Self, Fail> {
```

Comment:
Is there a reason not to use a &mut self rather than returning a new socket?

Reply:
Yes, there is, because we are not changing the state of the target socket, but rather constructing a new one from the target socket.